src/ecredis.erl

```diff
@@ -0,0 +1,420 @@
+    slots_maps :: tuple(), %% Tuple with all the Redis nodes and Redis slots mapped on them.
+    version :: integer()  %% Used to avoid unnecessary refresh of Redis slots.
+}).
+
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+% API
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+
+-spec start_link([]) -> {ok, pid()}.
+start_link({ClusterName, Args}) ->
+    gen_server:start_link({local, ClusterName}, ?MODULE, [ClusterName, Args], []).
+
+
+%% Looks up eredis Pid for Slot.
+-spec get_eredis_pid(ClusterName::atom(), Slot::integer()) -> 
+    {Pid::pid(), Version::integer()} | undefined.
+get_eredis_pid(ClusterName, Slot) ->
+    Result = ets:lookup(ets_table_name(ClusterName), Slot),
+    case Result of
+        [] -> undefined;
+        [{_, Result2}] -> Result2
+    end. 
+
+
+%% Remaps various redis slots.
+-spec remap_cluster(ClusterName::atom(), Version::integer()) -> {Version::integer()}.
+remap_cluster(ClusterName, Version) ->
+    Result = gen_server:call(ClusterName, {remap_cluster, Version}),
+    Result.
+
+
+%% Executes qp.
+-spec qp(ClusterName::atom(), redis_pipeline_command()) -> redis_pipeline_result().
+qp(ClusterName, Commands) ->
+    q(ClusterName, Commands).
+
+
+%% Executes q.
+-spec q(ClusterName::atom(), redis_command()) -> redis_result().
+q(ClusterName, Command) ->
+    query(ClusterName, Command).
+
+
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+% Internal methods
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+
+
+-spec connect_(State::#state{}, [{Address::string(), Port::integer()}]) -> #state{}.
+connect_(_State, []) ->
+    #state{};
+connect_(State, InitNodes) ->
+    NewState = State#state{
+        slots_cache = {},
+        slots_maps = {},
+        init_nodes = [#node{address = A, port = P} || {A,P} <- InitNodes],
+        version = 0
+    },
+    reload_slots_map(NewState).
+
+
+-spec remap_cluster_internal(State::#state{}, Version::integer()) -> State::#state{}.
+remap_cluster_internal(State, Version) ->
+    if State#state.version == Version ->
+        reload_slots_map(State);
+    true ->
+        State
+    end.
+
+
+%% TODO(vipin): Need to reset connection on Redis node removal.
+-spec reload_slots_map(State::#state{}) -> State::#state{}.
+reload_slots_map(State) ->
+    ClusterSlots = get_cluster_slots(State#state.cluster_name, State#state.init_nodes),
+    SlotsMaps = parse_cluster_slots(ClusterSlots),
+    ConnectedSlotsMaps = connect_all_slots(State#state.cluster_name, SlotsMaps),
+    SlotsCache = create_slots_cache(ConnectedSlotsMaps),
+
+    NewState = State#state{
+        slots_cache = list_to_tuple(SlotsCache),
+        slots_maps = list_to_tuple(ConnectedSlotsMaps),
+        version = State#state.version + 1
+    },
+    [cache_eredis_pids(NewState, Slot) || Slot <- lists:seq(0, ?REDIS_CLUSTER_HASH_SLOTS - 1)],
+
```

Comment:
I agree with nikola. this logic can be simplified here.
once we have the cluster slots and have connected to these nodes. we can directly call the logic to cache it: and send the slots map to the function and insert it into the ets table.